### DISCUSSION TITLE
Pattern checker: Download video or audio using cobalt.tools

### DISCUSSION THREAD

**Discussion by squeak9** [TrianguloY/URLCheck#386] [Show and tell]
https://cobalt.tools/ is a great website for downloading videos or mp3s from lots of website like youtube, reddit etc. It doesn't have paid features or limits or any of that stuff. 

It has a [url query feature](https://github.com/imputnet/cobalt/issues/234) which we can use with the pattern checker module to save us the pasting of links into the website:

```json
  "Try to download using cobalt.tools": {
    "regex1": "^(https?):\/\/(\\S*)$",
    "regex": "^(?!.*(cobalt.))(.*)",
    "replacement": "https:\/\/cobalt.tools\/#$2",
    "enabled": "true"
```
Note that this lets you use it for any website cause it supports a lot and more might get added: https://github.com/imputnet/cobalt#supported-services
If there's someone who's better at regex than me that can get it to only trigger on certain domains that would be amazing but as it is it should prove handy.

**Comment by TrianguloY** [comment]
Thanks for the pattern!
I want to add it to the wiki, but I think it can be simplified a bit.
First, the regex1 is unused, but I guess that's just an alternative pattern just in case. That one seems to match all `http://` and `https://` starting urls without spaces.
The pattern itself seems odd, as it will basically match anything that does not contains the word 'cobalt' (plus a character) anywhere in the url.

Maybe a better pattern would be this one?
```json
"Try to download using cobalt.tools": {
    "regex": "^https?:\/\/(?!cobalt\\.tools)",
    "replacement": "https:\/\/cobalt.tools\/#$0"
}
```

The regexp can be expanded to only support specific domains like this:
```regex
^https?:\/\/(www\\.bilibili\\.com|www\\.youtube\\.com|www\\.tumblr\\.com)\/
```
but since cobalt supports a lot of different services, and the list seems to grow, maybe it is a good idea to keep the 'all' regex and let the user manually decide to apply it or not.

**Reply by squeak9** [reply]
Oh yeah, i was copying some other thing and i can't remember why i did it like that. This is way better, thanks for the help!
Add it to the wiki like that I say.

**Reply by Bakr-Ali** [reply]
> First, the regex1 is unused, but I guess that's just an alternative pattern just in case. 

@TrianguloY can you please explain what is the regex1 field and how to use it? Does it have to be 1 or can we attach any number to it (regex234)? Can there be more than 1 numbered regex (regex3 + regex4) and how would that work? It seems this is not documented anywhere.

**Reply by TrianguloY** [reply]
> @TrianguloY can you please explain what is the regex1 field and how to use it? Does it have to be 1 or can we attach any number to it (regex234)? Can there be more than 1 numbered regex (regex3 + regex4) and how would that work? It seems this is not documented anywhere.

It's a "user field" if you want to call it that. Basically the app will not use that, so you can write whatever you want.

The explanation is that, due to how the configuration is provided and saved (as json) you can write whatever valid json you want, and the app will save it, but only those keys from the documentation will be used. This means that you can write any extra key/value, like for example "foo", "comment", "example", "potato", "regex1", "regex2", "regex123", etc; and the app will just not care (will not use any of those).

That's why I said that probably that regex1 was something the original user wrote for themselves, maybe as an alternative one for the future, that the app will save but never use.
